cores/esp8266/HardwareSerial.cpp

```diff
@@ -162,6 +170,12 @@ size_t HardwareSerial::readBytes(char* buffer, size_t size)
 
 #if !defined(NO_GLOBAL_INSTANCES) && !defined(NO_GLOBAL_SERIAL)
 HardwareSerial Serial(UART0);
+
+// Executed at end of loop() processing when > 0 bytes available in the Serial port
+void serialEventRun(void)
+{
+  if (serialEvent && Serial.available()) serialEvent();
```

Comment:
Single line if

Reply:
I'm curious here: if serialEventRun is null, does this whole thing get optimized away?

Reply:
No, because `serialEventRun` is a linker-set variable and the GCC compile stage doesn't know about it.  It's one of the minor things LTO would help with (but LTO breaks everything else we're using w/weak functions...)

Reply:
I just disassembled a test w/no user `serialEvent`:

The `if (serialEventRun())` in `main.cpp` *is* still present:
````
40201120:       000000                  ill
...
40201144:       201110                  or      a1, a1, a1
40201147:       fffbc5                  call0   40201104 <__loop_end>
    if (serialEventRun) {
4020114a:       fff521                  l32r    a2, 40201120 <__loop_end+0x1c>
4020114d:       428c                    beqz.n  a2, 40201155 <_ZL12loop_wrapperv+0x31>
        serialEventRun();
4020114f:       fff401                  l32r    a0, 40201120 <__loop_end+0x1c>
40201152:       0000c0                  callx0  a0
    esp_schedule();
````

However, the `serialEventRun` function isn't present at all, so you just load that `00000000` and jump around the call each main loop_wrapper cycle.